proposal/color-4-new-spaces-js.d.ts.md

```diff
@@ -0,0 +1,864 @@
+
+export type ChannelNameHSL = 'hue' | 'saturation' | 'lightness';
+
+export type ColorSpaceHWB = 'hwb';
+
+export type ChannelNameHWB = 'hue' | 'whiteness' | 'blackness';
+
+export type ColorSpaceLAB = 'lab' | 'oklab';
+
+export type ChannelNameLAB = 'lightness' | 'a' | 'b';
+
+export type ColorSpaceLCH = 'lch' | 'oklch';
+
+export type ChannelNameLCH = 'lightness' | 'chroma' | 'hue';
+
+export type ColorSpaceRGB =
+  | 'a98-rgb'
+  | 'display-p3'
+  | 'prophoto-rgb'
+  | 'rgb'
+  | 'srgb'
+  | 'srgb-linear';
+
+export type ChannelNameRGB = 'red' | 'green' | 'blue';
+
+export type ColorSpaceXYZ = 'xyz' | 'xyz-d50' | 'xyz-d65';
+
+export type ChannelNameXYZ = 'x' | 'y' | 'z';
+
+export type ChannelName =
+  | ChannelNameHSL
+  | ChannelNameHWB
+  | ChannelNameLAB
+  | ChannelNameLCH
+  | ChannelNameRGB
+  | ChannelNameXYZ;
+
+export type KnownColorSpace =
+  | ColorSpaceHSL
+  | ColorSpaceHWB
+  | ColorSpaceLAB
+  | ColorSpaceLCH
+  | ColorSpaceRGB
+  | ColorSpaceXYZ;
+
+export type PolarColorSpace = ColorSpaceHSL | ColorSpaceHWB | ColorSpaceLCH;
+
+export type RectangularColorSpace = Omit<KnownColorSpace, PolarColorSpace>;
+
+export type HueInterpolationMethod =
+  | 'decreasing'
+  | 'increasing'
+  | 'longer'
+  | 'shorter';
+```
+
+### New Color Functions
+
+```ts
+export class SassColor extends Value {
+```
+
+#### `space`
+
+Returns the name of [`internal`]'s space.
+
+[`internal`]: ../spec/js-api/value/color.d.ts.md#internal
+
+```ts
+get space(): KnownColorSpace;
+```
+
+#### `toSpace`
+
+Returns the result of [`color.to-space(internal, space)`] as a new SassColor.
+
+```ts
+toSpace(space: KnownColorSpace): SassColor;
+```
+
+[`color.to-space(internal, space)`]: ./color-4-new-spaces.md#colorto-space
+
+#### `isLegacy`
+
+Returns whether [`internal`] is in a [legacy color space] (`rgb`, `hsl`, or
+`hwb`).
+
+```ts
+get isLegacy(): boolean;
+```
+
+[legacy color space]: ./color-4-new-spaces.md#legacy-color
+
+#### `isInGamut`
+
+Returns the result of [`color.is-in-gamut(internal, space)`].
+
+```ts
+isInGamut(space?: KnownColorSpace): boolean;
+```
+
+[`color.is-in-gamut(internal, space)`]: ./color-4-new-spaces.md#coloris-in-gamut
+
+#### `toGamut`
+
+Returns the result of [`color.to-gamut(internal, space)`].
+
+```ts
+toGamut(space?: KnownColorSpace): SassColor;
+```
+
+[`color.to-gamut(internal, space)`]:  ./color-4-new-spaces.md#colorto-gamut-1
+
+#### `channelsOrNull`
+
+Returns an array of channel values for [`internal`], with [missing
+channels][missing components] converted to `null`.
+
+* Let `space` be the result of [`this.space`].
+
+* Let `components` be the list of channels in `space` and "alpha".
+
+* Let `channels` be an empty array.
+
+* For each `component` in `components`:
+
+  * Let `value` be the channel value in `color` with name of `component`.
+
+  * If `value` is `none`, let `value` be `null`.
+
+  * Append `value` to `channels`.
+
+* Return `channels`.
+
+```ts
+get channelsOrNull(): [
```

Comment:
I think this should only include the three channels that are different between color spaces.

Reply:
I think I misinterpreted an earlier comment about including alpha as a channel, but I think this makes more sense, given the channel vs component distinction. I removed alpha from `channelsOrNull` and `channels` in 03254fa2ff5902a4ae3fba6a854d06e680d0a379.